spec/acts_as_tag_pgarray/taggable_spec.rb

```diff
@@ -1,162 +1,275 @@
+    describe "#colors_cloud" do
+      it "returns tag cloud for tag_name" do
+        expect(User.colors_cloud).to match_array(
+          [@user1, @user2, @user3, @admin1, @admin2].map(&:colors).flatten.group_by(&:to_s).map { |k, v| [k, v.count] }
+        )
+      end
+
+      it "returns filtered tag cloud for tag_name with block" do
+        expect(User.colors_cloud { where(name: ["Ken", "Joe"]) }).to match_array(
+          [@user2, @user3].map(&:colors).flatten.group_by(&:to_s).map { |k, v| [k, v.count] }
+        )
+      end
+
+      it "returns filtered tag cloud for tag_name with prepended scope" do
+        expect(User.where("tag like ?", "bl%").colors_cloud).to match_array(
+          [@user1, @user2, @user3, @admin2].map(&:colors).flatten.group_by(&:to_s).map { |k, v| [k, v.count] }.select { |name, count| name.start_with? "bl" }
+        )
+      end
+
+      it "returns filtered tag cloud for tag_name with prepended scope and block" do
+        expect(User.where("tag like ?", "bl%").colors_cloud { where(name: ["Ken", "Joe"]) }).to match_array(
+          [@user2, @user3].map(&:colors).flatten.group_by(&:to_s).map { |k, v| [k, v.count] }.select { |name, count| name.start_with? "bl" }
+        )
+      end
     end
 
-    it "returns filtered tags for tag_name with prepended scope and bock" do
-      expect(User.where("tag like ?", "bl%").all_colors { where(name: ["Ken", "Joe"]) }).to match_array([@user2, @user3].map(&:colors).flatten.uniq.select { |name| name.start_with? "bl" })
+    describe "with complex scope" do
+      it "works properly" do
+        expect(User.without_any_colors("white").with_any_colors("blue").order(:created_at).limit(10)).to eq [@user1, @user3]
+      end
     end
   end
 
-  describe "#colors_cloud" do
-    it "returns tag cloud for tag_name" do
-      expect(User.colors_cloud).to match_array(
-        [@user1, @user2, @user3, @admin1, @admin2].map(&:colors).flatten.group_by(&:to_s).map { |k, v| [k, v.count] }
-      )
+  context "with allowed option" do
+    context "when options is not an array" do
+      it "raises an error" do
+        expect { User.acts_as_taggable_array_on :colors, allowed: "red, blue" }
+          .to raise_error(described_class::InvalidAllowListTypeError)
+      end
     end
 
-    it "returns filtered tag cloud for tag_name with block" do
-      expect(User.colors_cloud { where(name: ["Ken", "Joe"]) }).to match_array(
-        [@user2, @user3].map(&:colors).flatten.group_by(&:to_s).map { |k, v| [k, v.count] }
-      )
-    end
+    context "when options is an array" do
+      context "when tag is array of varchar type" do
+        shared_examples "allowed with varchar array" do
+          it "accepts allowed values as strings" do
+            expect(User.new(colors: %w[blue red])).to be_valid
+            expect(User.new(colors: %w[red])).to be_valid
```

Comment:
Your testing for array with symbols, string or integer values. But what if anything else will be provided? A hash like object inside array? Any other object? class name?

Reply:
I think that we could throw an exception in this case, what do you think?

Reply:
I think that we can go forward now supporting only these three kinds of types, integers, strings and symbols, do you agree?

Reply:
Let's start with just integers, strings, symbols inside an array. 

It's probably possible to check if any other object is serializable, but let's not go that far for now.